### ISSUE TITLE
Automatic Merge In Timeout

### ISSUE THREAD

**Issue by 0x4007** [ubiquibot/plugins-wishlist#20]
We need to maintain a bias of always moving forward in Ubiquity. 

This plugin can be considered dangerous if not configured properly, but the general idea is that if reviewers forget about a pull, it should automatically be merged in after a timeout period.

Default configs:

1. Minimum approvals: 1
2. Collaborator Merge Timeout (Days): 3.5
3. Contributor Merge Timeout (Days): -1[^1^]

Worst case scenario, the pull can be reverted. If a pull is detected to be reverted, perhaps a stretch goal of this plugin would be to disable automatic merges again for pulls associated to the same task. 

[^1^]: Represents disabled.

**Comment by 0x4007** [comment]
I assume it'll take a week to also write unit tests etc.

**Comment by gentlementlegen** [comment]
I would suggest a few changes:
The configuration can be made like
Collaborator Minimum Approvals Required: "0 days"
Contributor Minimum Approvals Required: "1 day"
Collaborator Merge Timeout (Days): "3.5 days"
Contributor Merge Timeout (Days): "7 days"

And this plugin should not need to be able to generate permits since that would be occurring within `conversation-rewards` on PR merging.

A condition that should be there before merge is that all Checks should be green as well, and of course no conflicts.

**Comment by rndquu** [comment]
> This plugin can be considered dangerous if not configured properly

This plugin is dangerous no matter how we configure it

> if reviewers forget about a pull

We need a dedicated QA engineer

> it should automatically be merged in after a timeout period

We mustn't allow auto merges of unreviewed and unQAed PRs

**Comment by gentlementlegen** [comment]
I agree with your points. Albeit I think that's already how we kinda manually do things, since we wait for one reviewer to go through at least before considering merging the pull request. For the QA part, this should be covered by the unit tests, for most cases.

I was thinking of eventually create a warning message before the automatic merge that tags the reviewers before the merge happens.

**Comment by rndquu** [comment]
> I agree with your points. Albeit I think that's already how we kinda manually do things, since we wait for one reviewer to go through at least before considering merging the pull request. For the QA part, this should be covered by the unit tests, for most cases.
> 
> I was thinking of eventually create a warning message before the automatic merge that tags the reviewers before the merge happens.

This "automerge plugin" is implied to solve an issue of long reviews but it does so by removing reviews at all thus reducing security and code quality levels. 

Long reviews should be solved either by a dedicated QA engineer either by PR comment incentives for outside collaborators.

Any unreviewed PR is basically a none 0 risk of exfiltrating any critical ubiquity organization secret (`X25519_PRIVATE_KEY`, `UBIQUIBOT_APP_PRIVATE_KEY`, `NPM_TOKEN`, etc...).

**Comment by gentlementlegen** [comment]
So wouldn't you consider it to be safe enough to if we:
- enforce at least one reviewer either case (collaborator, contributor)
- check for conflicts
- check for full pass on Action checks

**Comment by rndquu** [comment]
> enforce at least one reviewer either case (collaborator, contributor)

Sounds good

**Comment by 0x4007** [comment]
@gentlementlegen my formatting score isn't fully using the decimal library. Look at the tiny fraction leftover.

**Comment by gentlementlegen** [comment]
@0x4007 It is using it but one value must be used after it was already computed as a basic number. Adding a ticket.

**Comment by 0x4007** [comment]
The developer working on the sprint will be able to determine if they are related but I suppose we generally offer credit either way for multiple tasks not being redundant if the specification author doesn't know from the start.
